File: Civitas-Java/Civitas/src/main/java/civitas/CivitasJuego.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
import java.util.Collections;

/**
 * @author elena
 * @date 04/10/2019
 */

public class CivitasJuego {
    private int indiceJugador;
    private Tablero tablero;
    private MazoSorpresas mazo;
    private EstadosJuego estado;
    private GestorEstados gestorEstados;
    private ArrayList<Jugador> jugadores;

    public CivitasJuego(ArrayList<String> nombres){
        jugadores = new ArrayList();
        
        for(String n : nombres){
            jugadores.add(new Jugador(n));
        }
        
        gestorEstados.estadoInicial();
        
        indiceJugador = Dado.getInstance().tirar();
        mazo = new MazoSorpresas();
        this.inicializarTablero(mazo);
        this.inicializarMazoSorpresas(tablero);
    }
    
    private void inicializarTablero(MazoSorpresas mazo){
        tablero = new Tablero(5);
        
        tablero.añadeCasilla(new Casilla("Salida"));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Recogidas",
        50, 20, 100, 100, 50)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Tranvia",
        55, 20, 120, 150, 60)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Aeropuerto",
        60, 20, 140, 200, 70)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Sana",
        65, 20, 160, 250, 80)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Ole",
        80, 20, 200, 300, 90)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Mesa",
        85, 20, 220, 350, 100)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Trompeta",
        90, 20, 240, 400, 110)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Maracas",
        95, 20, 260, 450, 120)));
        tablero.añadeCasilla(new Casilla(400, "Impuesto"));//impuesto
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Solida",
        110, 20, 300, 500, 130)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Coche",
        115, 20, 320, 550, 140)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Pienso",
        120, 20, 340, 600, 150)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Tren",
        125, 20, 360, 650, 160)));
        tablero.añadeJuez();//juez
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Mal",
        140, 20, 400, 700, 170)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Guitarra",
        145, 20, 420, 750, 180)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Hora",
        155, 20, 440, 800, 190)));
        tablero.añadeCasilla(new Casilla(new TituloPropiedad("Calle Diamante",
        170, 20, 460, 850, 200)));        
    }
    
    private void inicializarMazoSorpresas(Tablero tablero){
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRCARCEL, tablero));
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRCARCEL, tablero));
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA, tablero, 4, "Ve a la casilla 4"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.IRCASILLA, tablero, 12, "Ve a la casiila 12"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PAGARCOBRAR, -500, "Paga a otro jugador"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PAGARCOBRAR, 500, "Recibes dinero *-*"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PORCASAHOTEL, 100, "Pagas por cada casa y hotel"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PORJUGADOR, 50, "Recibes dinero de los demas"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.SALIRCARCEL, mazo));
        mazo.alMazo(new Sorpresa(TipoSorpresa.SALIRCARCEL, mazo));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PORJUGADOR, -50, "Dale dinero a los demás"));
        mazo.alMazo(new Sorpresa(TipoSorpresa.PAGARCOBRAR, -100, "Pagas a la banca"));                           
    }
    
    private void contabilizarPasosPorSalida(Jugador jugadorActual){
        while(tablero.getPorSalida() > 0){
            jugadorActual.pasaPorSalida();
        }
    }
    
    private void pasarTurno(){ indiceJugador = (indiceJugador+1)%jugadores.size(); }
    
    public void siguientePasoCompletado(OperacionesJuego operacion){
        this.estado = gestorEstados.siguienteEstado(jugadores.get(indiceJugador), estado, operacion);
    }
    
    public boolean construirCasa(int ip){ return jugadores.get(indiceJugador).construirCasa(ip); }
    
    public boolean construirHotel(int ip){ return jugadores.get(indiceJugador).construirHotel(ip); }
    
    public boolean vender(int ip){ return jugadores.get(indiceJugador).vender(ip); }
      
    public boolean hipotecar(int ip){ return jugadores.get(indiceJugador).hipotecar(ip); }
    
    public boolean cancelarHipoteca(int ip){ return jugadores.get(indiceJugador).cancelarHipoteca(ip); }
    
    public boolean salirCarcelPagando(){ return jugadores.get(indiceJugador).salirCarcelPagando(); }
    
    public boolean salirCarcelTirando(){ return jugadores.get(indiceJugador).salirCarcelTirando(); }
    
    public boolean finalDelJuego(){ 
        for(Jugador j : jugadores){
            if(j.enBancarrota()) return true;
        }
        return false;
    }
    
    private ArrayList<Jugador> ranking(){ 
        Collections.sort(jugadores);
        return jugadores;
    }
    
    public Casilla getCasillaActual(){ return tablero.getCasilla(jugadores.get(indiceJugador).getNumCasillaActual()); }
    
    public Jugador getJugadorActual(){ return jugadores.get(indiceJugador); }    
    
    public String infoJugadorTexto(){ return jugadores.get(indiceJugador).toString(); }   
        
    private void avanzaJugador(){ throw new UnsupportedOperationException("No implementado"); }      
    public boolean comprar(){ throw new UnsupportedOperationException("No implementado"); }   
    public OperacionesJuego siguientePaso(){ throw new UnsupportedOperationException("No implementado"); }
    
}

File: Civitas-Ruby/pkg/Civitas-Ruby-0.0.1/lib/civitas_juego.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "tablero"
require_relative "operaciones_juego"
require_relative "mazo_sorpresas"
require_relative "dado"
require_relative "gestor_estados"
require_relative "estados_juego"
require_relative "jugador"

module Civitas
  class CivitasJuego
    def initialize(nombres)
      @jugadores = Array.new
      for n in nombres
        @jugadores << Jugador.new(n)
      end
      
      @gestorestados = Gestor_estados.new
      @estado =  @gestorestados.estado_inicial
      
      @indiceJugadorActual = Dado.instance.tirar
      
      @mazo = MazoSorpresas.new
      inicializartablero(@mazo)
      inicializarmazosorpresas(@tablero)
    end
    
    def cancelarhipoteca(ip)
      @jugadores[@indiceJugadorActual].cancelarhipoteca(ip)
    end
    
    def comprar      
    end
    
    def construircasa(ip)
      @jugadores[@indiceJugadorActual].construircasa(ip)
    end
    
    def construirhotel(ip)
      @jugadores[@indiceJugadorActual].construirhotel(ip)
    end
    
    def finaldeljuego
      for j in @jugadores
        if j.enbancarrota
          true
        end
      end
      false
    end
    
    def getcasillaactual
      @tablero.getcasilla(@jugadores[@indiceJugadorActual].getNumCasillaActual)
    end
    
    def getjugadoractual
      @jugadores[@indiceJugadorActual]
    end
    
    def hipotecar(ip)
      @jugadores[@indiceJugadorActual].hipotecar(ip)
    end
    
    def infojugadortexto
      @jugadores[@indiceJugadorActual].to_s
    end
    
    def salircarcelpagando
      @jugadores[@indiceJugadorActual].salircarcelpagando(ip)
    end
    
    def salircarceltirando
      @jugadores[@indiceJugadorActual].salircarceltirando(ip)
    end
    
    def siguientepaso      
    end
    
    def siguientepasocompletado(operacion)
      @estado = @gestorestados.siguiente_estado(@jugadores[@indiceJugadorActual], @estado, operacion)
    end
    
    def vender(ip)
      @jugadores[@indiceJugadorActual].vender(ip)
    end
    
    private
    def avanzajugador      
    end
    
    def contabilizarpasosporsalida(jugadoractual)
      if @tablero.PorSalida > 0
        jugadoractual.pasaporsalida
      end
    end
    
    def inicializarmazosorpresas(tablero)
      @mazo.almazo(Sorpresa.new_sorpresa_ircarcel(Civitas::TipoSorpresa::IRCARCEL, tablero))
      @mazo.almazo(Sorpresa.new_sorpresa_ircarcel(Civitas::TipoSorpresa::IRCARCEL, tablero))
      @mazo.almazo(Sorpresa.new_sorpresa_casilla(Civitas::TipoSorpresa::IRCASILLA, tablero, 4, 'Ve a la casilla 4'))
      @mazo.almazo(Sorpresa.new_sorpresa_casilla(Civitas::TipoSorpresa::IRCASILLA, tablero, 12, 'Ve a la casilla 12'))
      @mazo.almazo(Sorpresa.new_sorpresas(Civitas::TipoSorpresa::PAGARCOBRAR, -500, 'Paga a otro jugador'))
      @mazo.almazo(Sorpresa.new_sorpresas(Civitas::TipoSorpresa::PAGARCOBRAR, 500, 'Recibes dinero *-*'))
      @mazo.almazo(Sorpresa.new_sorpresas(Civitas::TipoSorpresa::PORCASAHOTEL, 100, 'Pagas por cada casa y hotel'))
      @mazo.almazo(Sorpresa.new_sorpresas(Civitas::TipoSorpresa::PORJUGADOR, 50, 'Recibes dinero de los demas'))
      @mazo.almazo(Sorpresa.new_sorpresas(Civitas::TipoSorpresa::PORJUGADOR, -50, 'Dale dinero a los demas'))
      @mazo.almazo(Sorpresa.new_sorpresa_evitarcarcel(Civitas::TipoSorpresa::SALIRCARCEL, @mazo))
      @mazo.almazo(Sorpresa.new_sorpresa_evitarcarcel(Civitas::TipoSorpresa::SALIRCARCEL, @mazo))
      @mazo.almazo(Sorpresa.new_sorpresa_evitarcarcel(Civitas::TipoSorpresa::SALIRCARCEL, @mazo))                          
    end
    
    def inicializartablero(mazo)
      @tablero = Tablero.new(5)
      
      @tablero.aniadecasilla(Casilla.new_casilla('Salida'))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Recogidas',
      50, 20, 100, 100, 50)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Tranvia',
      55, 20, 120, 150, 60)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Aeropuerto',
      60, 20, 140, 200, 70)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Sana',
      65, 20, 160, 250, 80)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Ole',
      80, 20, 200, 300, 90)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Mesa',
      85, 20, 220, 350, 100)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Trompeta',
      90, 20, 240, 400, 110)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Maracas',
      95, 20, 260, 450, 120)))
      @tablero.aniadecasilla(Casilla.new_casillaimpuesto(400, 'Impuesto'))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Solida',
      110, 20, 300, 500, 130)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Coche',
      115, 20, 320, 550, 140)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Pienso',
      120, 20, 340, 600, 150)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Tren',
      125, 20, 360, 650, 160)))
      @tablero.aniadejuez
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Mal',
      140, 20, 400, 700, 170)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Guitarra',
      145, 20, 420, 750, 180)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Hora',
      155, 20, 440, 800, 190)))
      @tablero.aniadecasilla(Casilla.new_casillacalle(TituloPropiedad.new('Calle Diamante',
      170, 20, 460, 850, 200))) 
    end
    
    def pasarturno
      @indiceJugadorActual = (@indiceJugadorActual+1)%@jugadores.length
    end
    
    def ranking
      @jugadores.sort
    end
  end
end
